src/chains/ethereum/ethereum/src/helpers/storage-range-at.ts

```diff
@@ -0,0 +1,152 @@
+import { Data, findInsertPosition } from "@ganache/utils";
+import { decode } from "@ganache/rlp";
+
+import type { LevelUp } from "levelup";
+import type {
+  StorageRangeAtResult,
+  StorageRecords
+} from "@ganache/ethereum-utils";
+import { BaseTrie as Trie } from "merkle-patricia-tree";
+
+type TrieData = { key: Buffer; value: Buffer };
+type HashedKeysWithValues = {
+  length: number;
+  values: Buffer[];
+  keys: Buffer[];
+  nextKey: Data | null;
+};
+
+const compare = (a: Buffer, b: Buffer) => a.compare(b) < 0;
+
+/**
+ * Iterates over all keys in the trie, drops keys that sort before `startKey`,
+ * sorts them, and then returns them in sorted order, along with their values.
+ *
+ * The `keys` and `values` fields may contain more than `maxKeys` records. The
+ * `length` property must be used determine which keys are relevant.
+ *
+ * The value property is the raw value, it has not been RLP decoded.
+ *
+ * `nextKey` is set if there are more than `maxKeys` after `startKey`.
+ *
+ * @param trie
+ * @param startKey
+ * @param maxKeys
+ * @returns
+ */
+async function getHashedKeysWithValues(
+  trie: Trie,
+  startKey: Buffer,
+  maxKeys: number
+) {
+  return await new Promise<HashedKeysWithValues>((resolve, reject) => {
+    const keys: Buffer[] = [];
+    const values: Buffer[] = [];
+
+    trie
+      .createReadStream()
+      .on("data", function onData({ key, value }: TrieData) {
+        // ignore anything that comes before our starting point
+        if (startKey.compare(key) > 0) return;
+
+        // insert the key exactly where it needs to go in the array
+        const position = findInsertPosition(keys, key, compare);
```

Comment:
totally unnecessary thought on optimizing.

If we hold `maxKey` in memory and `maxKey` === the highest key value in the `keys` array 
could we compare `keys.length === maxKeys && key > maxKey` return? This would guarantee we only find the insert position for keys that could be in the array once it is full but automatically dropping keys that are too high


I think that would guarantee you always have the keys you want but it would only insertion sort `maxKeys` instead of keys greater than `startKey` 

This _might_ eliminate the need for the `length > maxKey` in the `end`

Reply:
I originally thought you were on to something (and you still might be) and started implementing this, but I realized ( I think) that if `key` ends up being less than `highestKey`, we still will need to insert into the array, which will make the array too long; so we still need to trim it down to size somehow. So we don't eliminate the `length > maxKeys` check.

Additionally, we _need_ to collect one more key than `maxKeys`, in order to find and return the `nextKey` value. So while there may be something to your optimization idea, implementing it started getting a bit complex so I've bailed and am here writing this comment now. Let me know if you want to pair on this and make we can make the optimization make more sense!

Reply:
Much appreciation to the time put in. 

I [made a branch](https://github.com/trufflesuite/ganache/blob/storageRangePrune/src/packages/utils/tests/utils.test.ts) with unit tests to show existing vs suggested. 

The TLDR is [one line](https://github.com/trufflesuite/ganache/blob/storageRangePrune/src/packages/utils/tests/utils.test.ts#L101).

I believe the existing code is the 'corner office' (best) solution.

It _is_ possible for the suggested custom solution to work without the `if (position > maxKeys) return;` however performance suffers greatly. 

The only time I saw the linked suggestion bench faster is when there were >1000 keys but I believe this is only due to the test conditions -> I simplified it to test against integers. The custom should prune operations when `position` equals `maxLength-1`. E.g. a key collision for the final item in the results array. 

I don't believe that is possible in production. Sharing code/thoughts for code/thoughts sake.

Reply:
So would you like to chase this optimization further or is the code good as is?

Reply:
> I believe the existing code is the 'corner office' (best) solution.

Good to go